### ISSUE TITLE
Grad norm computation

### ISSUE THREAD

**Issue by belerico** [davda54/sam#26]
Hi there! I was watching your code and i have a doubt regarding the `_grad_nom(self)` method: from the SAM paper the grad norm used to compute epsilon_hat(w) defined by the following equation
![image](https://user-images.githubusercontent.com/18405289/120919337-5a3b0980-c6b9-11eb-8c5a-1da8dbf84d2e.png) supposing p=2 and then q=2 (1/p + 1/q = 1), would simply be the 2-norm of the gradients. So, shouldn't be the `_grad_norm(self)` method something like this:
```python
@torch.no_grad()
def _grad_norm(self):
    # put everything on the same device, in case of model parallelism
    shared_device = self.param_groups[0]["params"][0].device
    norm = torch.sqrt(
        sum(
            [
                torch.sum(
                    torch.square(
                        p.grad
                        * (torch.abs(p.grad) if group["adaptive"] else 1.0)
                    )
                ).to(shared_device)
                for group in self.param_groups
                for p in group["params"]
                if p.grad is not None
            ]
        ),
    )
    return norm
```
I've already tested it with SAM obtaining an accuracy of 97.17% and I'm now running tests with ASAM

**Comment by davda54** [comment]
Hi, this is exactly what `torch.norm(tensor, p=2)` is computing. See also this related issue: #16

**Comment by belerico** [comment]
Sorry, my bad! I didn't see the issue and the equality
